src/main/scala/io/citrine/lolo/linear/LinearRegression.scala

```diff
@@ -26,80 +28,77 @@ case class LinearRegressionLearner(
       }
 
-    /* If we are fitting the intercept, add a row of 1s */
-    val At = if (fitIntercept) {
-      new DenseMatrix(
-        indices.size + 1,
-        n,
-        trainingData.map(r => indices.map(r._1(_).asInstanceOf[Double]) :+ 1.0).flatten.toArray
-      )
-    } else {
-      new DenseMatrix(indices.size, n, trainingData.map(r => indices.map(r._1(_).asInstanceOf[Double])).flatten.toArray)
-    }
-    val k = At.rows
-
-    /* If the weights are specified, multiply At by them */
-    val weightsMatrix = weights.map(w => diag(new DenseVector(w.toArray)))
-    val Atw = if (weightsMatrix.isDefined) {
-      At * weightsMatrix.get
-    } else {
-      At
-    }
-    val A = Atw.t
+    val numSamples = trainingData.length
+    val numFeatures = indices.length + (if (fitIntercept) 1 else 0)
 
-    val b = if (weightsMatrix.isDefined) {
-      new DenseVector(trainingData.map(_._2.asInstanceOf[Double]).zip(weights.get).map(p => p._1 * p._2).toArray)
+    /* Assemble breeze vectors for solving normal equations  */
+    val Xt = if (fitIntercept) {
+      val featureArray = features.flatMap { row => indices.map(idx => row(idx).asInstanceOf[Double]) :+ 1.0 }.toArray
+      new DenseMatrix(numFeatures, numSamples, featureArray)
     } else {
-      new DenseVector(trainingData.map(_._2.asInstanceOf[Double]).toArray)
+      val featureArray = features.flatMap { row => indices.map(idx => row(idx).asInstanceOf[Double]) }.toArray
+      new DenseMatrix(numFeatures, numSamples, featureArray)
     }
+    val X = Xt.t
+    val y = new DenseVector(labels.map(_.asInstanceOf[Double]).toArray)
 
-    val beta = if (regParam.exists(_ > 0) || n >= k) {
-      /* Construct the regularized problem and solve it */
-      val regVector = Math.pow(regParam.getOrElse(0.0), 2) * DenseVector.ones[Double](k)
-      if (fitIntercept) regVector(-1) = 0.0
-      val M = At * A + diag(regVector)
-      try {
-        val Mi = pinv(M)
-        /* Backsub to get the coefficients */
-        Mi * At * b
-      } catch {
-        case x: Throwable =>
-          val mean = if (weightsMatrix.isDefined) sum(b) / weights.get.sum else sum(b) / b.length
-          val res = DenseVector.zeros[Double](k)
-          res(-1) = mean
-          res
+    /* Rescale data by weight matrix */
+    val weightsMatrix = weights.map(w => diag(new DenseVector(w.toArray)))
+    val Xw = weightsMatrix.map(W => W * X).getOrElse(X)
+    val yw = weightsMatrix.map(W => W * y).getOrElse(y)
+
+    val (coefficients, intercept) = Try {
+      // If regularized/overdetermined, attempt to solve the full normal equations
+      // If underdetermined, fall back to a least-squares solution that uses SVD/pseudo-inverse internally
+      val beta = if (regularized || numSamples >= numFeatures) {
+        val l = math.pow(lambda, 2) * DenseVector.ones[Double](numFeatures)
+        if (fitIntercept) l(-1) = 0.0
+        val A = Xt * Xw + diag(l)
+        val b = Xt * yw
```

Comment:
Coming back to your question on zoom: yes, it is correct to use the weighted observations.